Project-Hotpot/Project-Hotpot/Managers/CacheManager.swift

```diff
@@ -0,0 +1,128 @@
+//
+//  CacheManager.swift
+//  Project-Hotpot
+//
+//  Created by Harini Sundaram on 7/27/22.
+//
+
+import Foundation
+
+class CacheManager: NSObject {
+    
+    //level of layering allows me to easily make changes to Cache implementation later on
+    private var cache : Cache
+    //I believe conventions are the constants are all caps
+    let MIN_CACHE_LIMIT = 0
+    let SONGS_PER_GENRE = 5
+    
+    private static var sharedCacheManager: CacheManager = { return CacheManager() }()
+    
+    override private init() {
+        cache = Cache(genreKeys: SpotifyManager.shared().originalGenreSeeds)
+        super.init()
+        
+    }
+    // MARK: - Accessors
+    class func shared() -> CacheManager {
+        return sharedCacheManager
+    }
+    
+    //TODO: unsure of naming in this case
+    func retrieveSongsFromCache(genre:String, completion: @escaping (_ result: Result<[SongDetails], Error>) -> Void) {
+        let songDetailsArray = cache.retrieveSongs(genre: genre)
+        if songDetailsArray == nil{
+            NSLog("unable to retreive from genre")
+            completion(.failure(CustomError.invalidCacheKey))
+        }
+        else if let songDetailsArray = songDetailsArray {
+            if songDetailsArray.count <= MIN_CACHE_LIMIT {
+                NSLog("Cache Miss")
+                refillCache(genre:genre) { result in
+                    switch result {
+                    case .success(let newSongDetailsArray):
+                        NSLog("refilled cache successfully")
+                        completion(.success(newSongDetailsArray))
+                    case .failure(let error):
+                        completion(.failure(error))
+                    }
+                }
+            }
+            else {
+                NSLog("Cache Hit")
+                completion(.success(songDetailsArray))
+            }
+        }
+    }
+    func removeSongFromCache(genre: String, song: SongDetails) {
+        cache.evictFromCache(genre: genre, evictSong: song)
+    }
+    func refillCache(genre:String, completion: @escaping (_ result: Result<[SongDetails], Error>) -> Void) {
+        fetchNSongs(limit:SONGS_PER_GENRE, genre:genre) { result in
+            switch result {
+            case .failure(let error):
+                completion(.failure(error))
+            case .success(let dictionary):
+                guard let tracks = dictionary["tracks"] as? [String:Any]?,
+                      let items = tracks?["items"] as? [[String:Any]?]
+                else{
+                    NSLog("failed parsing random song dictionary response")
+                    return completion(.failure(CustomError.failedResponseParsing))
+                }
+                let trackIDs : [String] = items.compactMap{ item in item?["id"] as? String }
```

Comment:
A little syntactic sugar; if you find yourself writing short map snippets that are essentially 1-liners, you can use `$0` to represent the passed in object, e.g. `{ $0["id"] as? String }`

("Shorthand Argument Names" in https://docs.swift.org/swift-book/LanguageGuide/Closures.html)